aiokafka/producer.py

```diff
@@ -496,3 +496,39 @@ def _partition(self, topic, partition, key, value,
         available = list(self._metadata.available_partitions_for_topic(topic))
         return self._partitioner(
             serialized_key, all_partitions, available)
+
+    def create_batch(self, topic, partition=None):
+        """Create and return an empty MessageBatch.
+
+        The batch is not queued for send until submission to ``send_batch``.
+        Use of both ``create_batch`` and ``send`` together is discouraged as it
```

Comment:
Just how are they unexpected? We should set a clear limit or allow both to be used.

Reply:
As I said in my previous PR, this is why I originally separated the classes out: to make it clearer that these methods aren't very compatible.

They don't play nicely because `send` ends up appending a message to the first batch with space--creating one if it doesn't exist. When a batch is sent via this `send_batch`, it's simply added to the `MessageAccumulator` queue. If that batch isn't full, calls to `send` that end up at the same TP will be appended to that batch, meaning that offsets won't match what the original caller was doing.

Another potential interaction comes with calls in this order:
1. send -- a fresh batch is created and appended to the queue
2. send_batch -- a full batch is added to the queue
3. send -- now waits because the full batch is the "active" one, even though plenty of space remains in the batch created in step 1

There are a couple other edge conditions but these are the big ones.

Reply:
> When a batch is sent via this send_batch, it's simply appended to the queue. If that batch isn't full, calls to send against the same TP will be appended to that batch, meaning that offsets won't match what the original batch caller was doing.

1) You need to close the batch for append it's submitted to `send_batch`. Ie. _is_full should return False.
2) Don't see why offsets would not match, as they are tracked in batch itself.

> Another potential interaction comes with calls in this order:

Maybe lets an explicit flag, when creating batches, that blocks `send` API until that batch is sent. Ie, every `create_batch` must match with `send_batch` calls. What do you think?

Reply:
> 2. Don't see why offsets would not match, as they are tracked in batch itself.

In that case, a message would have been added to the batch without the caller's knowledge so their expectation for what was written is no longer accurate. Whether this is a problem depends on what the caller's doing, but it would violate the explicit control I'm looking for here.

> Maybe lets an explicit flag, when creating batches, that blocks send API until that batch is sent. Ie, every create_batch must match with send_batch calls. What do you think?

Sure.